File: 377.combination-sum-iv.py
#
# @lc app=leetcode id=377 lang=python3
#
# [377] Combination Sum IV
#

# Time:  O()
# Space: O()

from typing import List


# @lc code=start
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [1] + [0] * target
        nums.sort()
        for i in range(1, target + 1):
            for j in range(len(nums)):
                if nums[j] <= i:
                    dp[i] += dp[i - nums[j]]
                else:
                    break
        return dp[target]


# @lc code=end

File: 377.combination-sum-iv.js
/*
 * @lc app=leetcode id=377 lang=javascript
 *
 * [377] Combination Sum IV
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var combinationSum4 = function (nums, target) {
  nums.sort((a, b) => a - b);
  const dp = Array(target + 1).fill(0);
  dp[0] = 1;
  for (let i = 1; i < target + 1; ++i) {
    for (let j = 0; j < nums.length; ++j) {
      if (nums[j] <= i) {
        dp[i] += dp[i - nums[j]];
      } else {
        break;
      }
    }
  }
  return dp[target];
};
// @lc code=end
